File: scala/src-scala/main/scala/domain/PokerEvents.scala
package domain

import akka.actor.ActorRef

case class PokerMessage(sender: String, text: String)

object SystemMessage {
  def apply(text: String) = PokerMessage("System", text)
}

sealed trait PokerEvent

case class IncomingMessage(sender: String,
                           message: String) extends PokerEvent

case class UserJoined(userName: String,
                      userActor: ActorRef,
                      isSpectator: Boolean = false,
                      eventType: String = "userJoined") extends PokerEvent

case class UserLeft(userName: String,
                    eventType: String = "userLeft") extends PokerEvent

case class RequestStartEstimation(userName: String,
                                  taskName: String,
                                  startDate: String,
                                  eventType: String = "startEstimation") extends PokerEvent

case class UserEstimate(userName: String,
                        taskName: String,
                        estimate: String,
                        eventType: String = "estimate") extends PokerEvent

case class UserHasEstimated(userName: String,
                            taskName: String,
                            eventType: String = "userHasEstimated") extends PokerEvent

case class RequestShowEstimationResult(userName: String,
                                       eventType: String = "showResult") extends PokerEvent

case class EstimationResult(taskName: String,
                            startDate: String,
                            endDate: String,
                            estimates: List[UserEstimation],
                            eventType: String = "estimationResult") extends PokerEvent

case class UserEstimation(userName: String, estimate: String) extends PokerEvent

case class HeartBeat(eventType: String = "keepAlive") extends PokerEvent

File: packages/backend/src/domain/pokerEvents.ts
type PokerEvent =
  | JoinRoom
  | UserJoined
  | UserLeft
  | ChangeCardDeck
  | StartEstimation
  | UserEstimate
  | UserHasEstimated
  | RequestShowEstimationResult
  | EstimationResult;

interface JoinRoom {
  eventType: 'joinRoom';
  userName: string;
  roomName: string;
  isSpectator: boolean;
}

interface UserJoined {
  eventType: 'userJoined';
  userName: string;
  isSpectator: boolean;
}

interface UserLeft {
  eventType: 'userLeft';
  userName: string;
}

interface ChangeCardDeck {
  eventType: 'changeCardDeck';
  cardDeck: string[];
}

interface StartEstimation {
  eventType: 'startEstimation';
  userName?: string;
  taskName: string;
  startDate: string;
}

interface UserEstimate {
  eventType: 'estimate';
  userName: string;
  taskName: string;
  estimate: string;
}

interface UserHasEstimated {
  eventType: 'userHasEstimated';
  userName: string;
  taskName: string;
}

interface RequestShowEstimationResult {
  eventType: 'showResult';
  userName: string;
}

interface EstimationResult {
  eventType: 'estimationResult';
  taskName: string;
  startDate: string;
  endDate: string;
  estimates: { userName: string; estimate: string }[];
}
